docs/design.md

```diff
@@ -0,0 +1,279 @@
+will trigger a new ingestion session to gather the data for that update. 
+- `Immediate` Ingestion: Some data providers may want to share updates for a small number of CIDs which doesn't
+require the generation of a dedicated `Advertisement` and triggering an ingestion session. For this case, indexer
+nodes will provide an endpoint for providers to push small updates directly to them. Indexer nodes may
+restrict immediate ingestion updates to authenticated data providers __(we3.storage case)__.
+
+## Ingestion Protocol
+When indexer nodes realize that they are not in sync with a data provider they are indexing data for, they trigger
+new ingestion sessions with them to gather the new updates. This is orchestrated by the ingestion protocol.
+
+Indexer nodes track locally the latest `AdvertisementID` seen for each data provider. When an `Advertisement` for an unseen `AdvertisementID` is seen for a data provider, a new ingestion session is triggered with the data provider.
+
+Data providers are responsible for broadcasting `Advertisement`s and generating their `AdvertisementID`s in the way that
+suits them best. This means that they may choose to use IPLD data structures under the hood to link their advertisements
+with their indexed data, so that `AdvertisementID` is a link to the indexed data for that `Advertisement`. If this is
+the case, the provider may set the `GraphSupport` flag in its `Advertisement` to signal indexer nodes that they
+can use IPLD-aware data transfer protocol, such as Graphsync, for the ingestion session.
+
+When an indexer node triggers a new `Sync` with provider `p`, for `latest AdvertisementID`:
+- Indexer node checks the last `AdvertisementID` seen for `p`, `current`.
+- It sends a request to ingest the data related with `latest` advertisement, `Sync(ctx, p, latest)`.
+
+According to how the data providers choose to represent their indexed data locally, two flavors of the ingestion protocols can be instantiated by the indexer node.
+
+### IPLD-aware Ingestion
+Initially, this will be default transport. If a data provider is indexing data in a way that `AdvertisementID` is linked with the indexed data using IPLD, the `GraphSupport` flag is enabled in its Advertisements to signal indexer nodes that they can use IPLD-aware data transfer
+protocols.
+
+When an indexer node sees an unseen advertisement `latest`, it will trigger an ingestion session to sync with the
+data provider. If `GraphSupport` is enabled, the indexer node can use Graphsync or `go-data-transfer` to fetch `latest` DAG, which links to the data that needs to be ingested, using a [`ExploreRecursiveEdge`](https://ipld.io/specs/selectors/) selector. `latest` will include a link to the data structure for the previous `AdvertisementID` to latest, so Graphsync is able to fetch all the chain of updates from the provider seamlessly up to the latest one seen by the indexer node.
+
```

Comment:
Will the new dag ever contain CIDs that the indexer already knows about and would be fetching unnecessarily?

Does the indexer need to know how to generate the difference between the two dags so that it knows which CIDs to remove?

Reply:
I think it is up to the provider to decide how it wants to link advertisements to entries, and how to provide the entries for ingestion.

For the Filecoin case this is quite straightforward, because Advertisements will probably be linked to new or expired deals, and deals that share CIDs will represent a metadata update in the index. 

I would expect duplicate CIDs on independent advertisements. The indexer should just disregard those that already has (which is what it already does when calling `Put` of `indexer-core`). 

On a related note, something we will need to figure out is how to perform snapshots of provider indexers to prune chains that are becoming too long. I'll add a placeholder in the doc.

Reply:
> I would expect duplicate CIDs on independent advertisements. The indexer should just disregard those that already has

Even if the CIDs are the same as previously seen, since they are part of a new deal they will almost certainly have different metadata.  This means that that the indexer will store multiple values for the same CID and provider.

I guess this is OK, and even good in that it allows the providers an easy way to delete CIDs that are no longer used.  When a deal expires, the provider tells the indexer to delete all the CIDs that have the metadata for that deal. If the CIDs were used in other deals, then the indexer will still have those values for the CIDs. The CID will only disappear from the indexer when it has no more values.

If the above sounds correct, then I will describe that in the design document. It should also be described clearly here to inform providers how to delete stale CIDs from the indexer without having to maintain state info (ref counts) for every CID in all their deals.